core/src/main/scala/org/ensime/core/Indexer.scala

```diff
@@ -24,10 +24,26 @@ class Indexer(
     LineSourcePositionHelper.fromFqnSymbol(hit)(vfs)
   )
 
-  def oldSearchTypes(query: String, max: Int) =
-    index.searchClasses(query, max).filterNot {
-      name => name.fqn.endsWith("$") || name.fqn.endsWith("$class")
-    }.map(typeResult)
+  def oldSearchTypes(query: String, max: Int) = {
+    val scalaNames = scala.collection.mutable.Set.empty[String]
+
+    def isUnique(fqn: String): Boolean = {
+      val stripped =
+        if (fqn.endsWith("$")) fqn.dropRight(1)
+        else if (fqn.endsWith("$class")) fqn.dropRight(6)
+        else fqn
+
+      if (scalaNames.contains(stripped)) {
+        false
+      } else {
+        scalaNames.add(stripped)
+        true
+      }
+    }
+
+    index.searchClasses(query, max)
```

Comment:
Ufff... I thought I missed something obvious, but now I see the misunderstanding. `name` is not case class and as far as I know there is no macro which defines `copy` for traits. This leads us to my previous post where I complained about converting trait to case class. I could do that with reflection, but I do not want to (I already broke most of FP good practices :D) Can it be done via new graph stuff?